libs/datamodel/core/src/transform/ast_to_dml/db/walkers.rs

```diff
@@ -0,0 +1,197 @@
+    pub(super) model_id: ast::ModelId,
+    pub(super) db: &'db ParserDatabase<'ast>,
+    pub(super) model_attributes: &'db ModelData<'ast>,
+}
+
+impl<'ast, 'db> ModelWalker<'ast, 'db> {
+    pub(crate) fn attributes(&self) -> &'db ModelData<'ast> {
+        self.model_attributes
+    }
+
+    #[allow(clippy::unnecessary_lazy_evaluations)] // respectfully disagree
+    fn final_database_name(&self) -> &'ast str {
+        self.model_attributes
+            .mapped_name
+            .unwrap_or_else(|| &self.db.ast[self.model_id].name.name)
+    }
+
+    #[allow(clippy::unnecessary_lazy_evaluations)] // respectfully disagree
+    fn get_field_db_names<'a>(&'a self, fields: &'a [ast::FieldId]) -> impl Iterator<Item = &'ast str> + 'a {
+        fields.iter().map(move |&field_id| {
+            self.db.types.scalar_fields[&(self.model_id, field_id)]
+                .mapped_name
+                .unwrap_or_else(|| &self.db.ast[self.model_id][field_id].name.name)
+        })
+    }
+
+    pub(crate) fn primary_key(&self) -> Option<PrimaryKeyWalker<'ast, 'db>> {
+        self.model_attributes.primary_key.as_ref().map(|pk| PrimaryKeyWalker {
+            model_id: self.model_id,
+            attribute: pk,
+            db: self.db,
+        })
+    }
+
+    pub(crate) fn walk_indexes(&self) -> impl Iterator<Item = IndexWalker<'ast, 'db>> + 'db {
+        let model_id = self.model_id;
+        let db = self.db;
+
+        self.model_attributes
+            .indexes
+            .iter()
+            .map(move |(index, index_attribute)| IndexWalker {
+                model_id,
+                index,
+                db,
+                index_attribute,
+            })
+    }
+
+    pub(crate) fn walk_relation_fields(&self) -> impl Iterator<Item = RelationFieldWalker<'ast, 'db>> + 'db {
+        let model_id = self.model_id;
+        let db = self.db;
+
+        self.db
+            .iter_model_relation_fields(self.model_id)
+            .map(move |(field_id, relation_field)| RelationFieldWalker {
+                model_id,
+                field_id,
+                db,
+                relation_field,
+            })
+    }
+}
+
+pub(crate) struct IndexWalker<'ast, 'db> {
+    model_id: ast::ModelId,
+    index: &'ast ast::Attribute,
+    db: &'db ParserDatabase<'ast>,
+    index_attribute: &'db super::types::IndexData<'ast>,
+}
+
+impl<'ast, 'db> IndexWalker<'ast, 'db> {
+    pub(crate) fn ast_attribute(&self) -> &'ast ast::Attribute {
+        self.index
+    }
+
+    pub(crate) fn final_database_name(&self) -> Cow<'ast, str> {
+        if let Some(mapped_name) = &self.index_attribute.db_name {
+            return Cow::Borrowed(mapped_name);
+        }
+
+        let model = self.db.walk_model(self.model_id);
+        let model_db_name = model.final_database_name();
+        let field_db_names: Vec<&str> = model.get_field_db_names(&self.index_attribute.fields).collect();
+
+        ConstraintNames::index_name(
```

Comment:
Do we actually need this `ConstraintNames` struct at all?

Reply:
nope, it's used more as a namespace. I think we should remove the struct to avoid it becoming stateful in the future, maybe in a next PR.